lib/Dialect/AIE/Transforms/AIEAssignLockIDs.cpp

```diff
@@ -40,50 +39,69 @@ struct AIEAssignLockIDsPass : AIEAssignLockIDsBase<AIEAssignLockIDsPass> {
 
-    // The second pass scans for locks with no lockIDs and assigns locks.
-    for (auto lock : device.getOps<LockOp>()) {
-      Operation *lock_tile = lock.getTile().getDefiningOp();
-      if (!lock.getLockID().has_value()) {
-        if (tileToLastID.find(lock_tile) == tileToLastID.end()) {
-          // If the tile operation corresponding to the lock does not exist in
-          // the data structure, initialize the lockID with 0 with an empty set.
-          tileToLastID[lock_tile].first = 0;
-        } else if (tileToLastID[lock_tile].first < 15) {
-          // If the tile operation of the lock exists, the potential lockID is
-          // checked with the set containing occupied lockIDs until a lockID
-          // that is free is found.
-          int potential_ID = tileToLastID[lock_tile].first;
-          while (true) {
-            if (tileToLastID[lock_tile].second.find(potential_ID) !=
-                tileToLastID[lock_tile].second.end())
-              potential_ID++;
-            else
-              break;
+    // Construct data structure storing locks by tile.
+    for (auto lockOp : device.getOps<LockOp>()) {
+
+      auto tileOp = lockOp.getTileOp();
+      bool isAssigned = lockOp.getLockID().has_value();
+
+      if (isAssigned) {
+        auto lockID = lockOp.getLockID().value();
+        auto iter = tileToLocks.find(tileOp);
+        if (iter == tileToLocks.end()) {
+          tileToLocks.insert({tileOp, {{lockID}, /* unassigned = */ {}}});
+        } else {
+          if (iter->second.assigned.find(lockID) !=
+              iter->second.assigned.end()) {
+            auto diag = lockOp->emitOpError("is assigned to the same lock (")
+                        << lockID << ") as another op.";
+            diag.attachNote(tileOp.getLoc())
+                << "tile has lock ops assigned to same lock.";
+            return signalPassFailure();
           }
-          tileToLastID[lock_tile].first = potential_ID;
+          iter->second.assigned.insert(lockID);
+        }
+      }
+
+      else {
+        auto iter = tileToLocks.find(tileOp);
+        if (iter == tileToLocks.end()) {
+          tileToLocks.insert({tileOp, {/* assigned = */ {}, {lockOp}}});
         } else {
-          lock->emitError() << "Exceeded the number of unique LockIDs";
-          return;
+          iter->second.unassigned.push_back(lockOp);
         }
+      }
+    }
 
-        // The lockID is assigned and is stored in the set.
-        lock->setAttr("lockID", rewriter.getI32IntegerAttr(
-                                    tileToLastID[lock_tile].first));
-        tileToLastID[lock_tile].second.insert(tileToLastID[lock_tile].first);
+    // IR mutation: assign locks to all unassigned lock ops.
+    for (auto &&[tileOp, locks] : tileToLocks) {
+
+      const auto locksPerTile =
+          getTargetModel(tileOp).getNumLocks(tileOp.getCol(), tileOp.getRow());
+
+      uint32_t nextID = 0;
+      for (auto &&lockOp : locks.unassigned) {
```

Comment:
Same here - not sure but do you need the `&&`?

Reply:
This will be deduced `const LockOp & lockOp`. If I had to write it explicitly I would drop the const-ref so will drop the && here too.